feat: Classify functions into categories and search functionality

sanjibansg opened the pull request:
This PR modifies the function index page to have structured categories for various functions and adds search functionality.

This should resolve #40

richtia reviewed (changes requested):
Thanks for making these updates!  

Here are a few suggestion for the overall site that would be nice.
* Would it be possible to have the search functionality prioritize function names before looking through the description test message?  For example if you search for `or`, it'll show lots of functions that have that word first.
* Search results don't go away when you click on other function categories.  Maybe it makes more sense to clear the search if a different function category on the left pane or home is clicked on.
* Seems like the home button doesn't really work anymore after you've made a search?


This might make sense to have in a separate PR:
* Can the actual function site redirection be displayed in the same scheme as the index?  Instead of it looking like it's taking you to a different website entirely?

richtia reviewed (changes requested):
Thanks for making the fixes.  I just noticed is that the search results have the function category prepended to the function name.  I think it came as part of my change previously to allow for duplicate function names to be displayed.  Can we fix those search results to only have the function name.  The category is displaying right over it, so it's not needed.  For some reason comparison functions like `and` and `or` can't be searched for properly.  They're likely some type of reserved word, can you look into that?

sanjibansg commented:
> Thanks for making the fixes. I just noticed is that the search results have the function category prepended to the function name. I think it came as part of my change previously to allow for duplicate function names to be displayed. Can we fix those search results to only have the function name. The category is displaying right over it, so it's not needed. For some reason comparison functions like `and` and `or` can't be searched for properly. They're likely some type of reserved word, can you look into that?

Fixed the function category being prepended to the names in the search results. This should improve the search results as well. Could you please confirm?

richtia commented:
> > Thanks for making the fixes. I just noticed is that the search results have the function category prepended to the function name. I think it came as part of my change previously to allow for duplicate function names to be displayed. Can we fix those search results to only have the function name. The category is displaying right over it, so it's not needed. For some reason comparison functions like `and` and `or` can't be searched for properly. They're likely some type of reserved word, can you look into that?
> 
> Fixed the function category being prepended to the names in the search results. This should improve the search results as well. Could you please confirm?

Looks like the update you made removed the prepended category from the function name in the search results.  But it also removed the prepended part from the resulting URL.  So it redirects to a page that doesn't exist.  We don't want to see the prepended category in the search result, but it's still needed in the page redirect.

sanjibansg commented:
> > > Thanks for making the fixes. I just noticed is that the search results have the function category prepended to the function name. I think it came as part of my change previously to allow for duplicate function names to be displayed. Can we fix those search results to only have the function name. The category is displaying right over it, so it's not needed. For some reason comparison functions like `and` and `or` can't be searched for properly. They're likely some type of reserved word, can you look into that?
> > 
> > 
> > Fixed the function category being prepended to the names in the search results. This should improve the search results as well. Could you please confirm?
> 
> Looks like the update you made removed the prepended category from the function name in the search results. But it also removed the prepended part from the resulting URL. So it redirects to a page that doesn't exist. We don't want to see the prepended category in the search result, but it's still needed in the page redirect.

ah yes, thanks for the catch! fixed it.

richtia commented on build_site.py:
```diff
@@ -4,16 +4,35 @@
 
 from bft.html.builder import build_site
 
+def copy_with_progress(src, dst, copy_function=shutil.copy2):
+    for root, dirs, files in os.walk(src):
+        relative_path = os.path.relpath(root, src)
+        destination_path = os.path.join(dst, relative_path)
+
+        os.makedirs(destination_path, exist_ok=True)
+
+        for file in files:
+            source_file = os.path.join(root, file)
+            destination_file = os.path.join(destination_path, file)
+            copy_function(source_file, destination_file)
+            print(f"Copying: {source_file} -> {destination_file}")
```
Let's use pathlib instead of os.path here and in the rest of this file.

You can use this as reference: https://docs.python.org/3/library/pathlib.html#correspondence-to-tools-in-the-os-module

sanjibansg replied:
made the change, thanks for the suggestion!
